sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala

```diff
@@ -154,14 +154,18 @@ private[hive] case class HiveGenericUDF(
     val refEvaluator = ctx.addReferenceObj("evaluator", evaluator)
     val evals = children.map(_.genCode(ctx))
 
+    val funcArg = ctx.freshName("funcArg")
+    val finalEval = ctx.freshName("finalEval")
     val setValues = evals.zipWithIndex.map {
       case (eval, i) =>
         s"""
-           |if (${eval.isNull}) {
-           |  $refEvaluator.setArg($i, null);
-           |} else {
-           |  $refEvaluator.setArg($i, ${eval.value});
-           |}
+           |final Object $finalEval = ${eval.isNull} ? null: ${eval.value};
+           |scala.Function0<Object> $funcArg = new scala.Function0<Object>() {
+           |    public Object apply() {
+           |      return $finalEval;
+           |    }
+           |  };
+           |$refEvaluator.setFuncArg($i, $funcArg);
```

Comment:
This won't work as we should mv the `eval.code` in apply, so that we can lazy run child functions. Just like the following code.
```
|$refEvaluator.setArg($i, new scala.runtime.AbstractFunction0<Object>() {
          |  public Object apply() {
          |    ${eval.code}
          |    return ${eval.value};
          |  }
          |});
          """.stripMargin
```